### ISSUE TITLE
Worse confidence scores on small Objects detect 

### ISSUE THREAD

**Issue by PushpakBhoge** [ultralytics/yolov5#8906]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi 

I am trying to train a model on custom dataset, the dataset has only one class and very small objects. To give you an idea at any time on a image of 1200x1200 highest object size is 64x64 most of the boxes are below 32x32

I have tiled my dataset with patch size of 512x512 using SAHI crop I did this to preserve the resolution which is necessary for small Objects

The dataset is huge around 100k images. after tiling there are 75% images with atleast one object and 25% negative samples. I am also confident about the quality of annotations error rate in annotations is less than 5%

I have trained it with different set of params and long schedules the mAP is good, bounding boxes are also on spot most of the times

Problem I am facing is all the predictions have very low confidence score I am talking about 0.05 hardly any bbox has score higher than 0.25, anything higher than 0.1 is rare

in training as well I evolved params over 60 generations on a small subset of data to get best hyper parameters, Object loss is not decreasing. At epoch 1 bbox  loss was 0.13 and object loss was 0.10 after 50 iterations bbox loss went down to 0.068 and mAP0.5 went up to 28. I have used pretrained v6 weights as a starting point, haven't tried training from scratch.q

But the object loss didn't budge, lowest I have seen is 0.998 that's it

What could be going wrong here ?

**Comment by glenn-jocher** [comment]
@PushpakBhoge 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. Low confidence scores are indicative of too short trainings and can be resolved by training for more epochs.

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by PushpakBhoge** [comment]
@glenn-jocher  Thank you for responding 
So the low confidence does not have anything to do with the extreme small size of objects right ?
I have trained the model for around 100 epoch should how long should I train beyond this point ? I have not seen any drop in obj loss over this epochs and other losses / metrics are stabilized
I am attaching training details here

This run was done using most of the settings at default probably cutmix, copypaste were set to 0.0 and mosaic set to 0.3 rest was left untouched. also, would you recommend training using a pretrained as a starting point or I would get better results using training from scratch ?

labels.jpg
![labels](https://user-images.githubusercontent.com/47291048/183696464-7fe5c2e6-a248-44dc-b93f-0a3413dc478c.jpg)
labels_correlogram.jpg
![labels_correlogram](https://user-images.githubusercontent.com/47291048/183695143-0f221e7a-4eed-4c88-8ada-5bea699dc54d.jpg)
P curve
![P_curve](https://user-images.githubusercontent.com/47291048/183695247-e33f1d3b-70ad-400b-aa76-f633221d049d.png)
R curve
![R_curve](https://user-images.githubusercontent.com/47291048/183695380-d7e5df0e-0fd2-4d0a-99c6-3fbc9eccc2ff.png)
pr curve
![PR_curve](https://user-images.githubusercontent.com/47291048/183695428-e5f204e4-fa25-4dff-90f6-3606ac3f386a.png)
f1 curve
![F1_curve](https://user-images.githubusercontent.com/47291048/183695597-f94796fb-ded0-4f95-a493-6fedb27bd66f.png)
results.png
![results](https://user-images.githubusercontent.com/47291048/183695707-ffff69b5-a4bb-4dd6-b0d0-3d470f8bfd0f.png)
confusion matrix
![confusion_matrix](https://user-images.githubusercontent.com/47291048/183695814-e0ca726d-6936-46b4-b6c0-fb96793154ac.png)

**Comment by glenn-jocher** [comment]
@PushpakBhoge for very small objects you'll benefit most from a larger train/detect --imgsz, i.e. 1280 or higher.

**Comment by PushpakBhoge** [comment]
yes that's why I am preserving the resolution of the original image I am not resizing image in any way. I am taking crops of 512x512 from original images using SAHI, each image provides 6-12 crops I basically tiled all the dataset like this and doing same thing at inference as well

**Comment by glenn-jocher** [comment]
@PushpakBhoge increased image size helps small objects even if expanded beyond native size. COCO performs better at 1280 even though no image is larger than 640.

Alternatively you could try a P2 model for small objects that goes down to stride 4:
https://github.com/ultralytics/yolov5/blob/master/models/hub/yolov5-p2.yaml

**Comment by PushpakBhoge** [comment]
hmmm so basically due to the small size the actual area of the object is small which may not be enough for the object loss to converge,
if we increased the resolution beyond native for example 512x512 to 1280x1280 the same object will be enlarged and the network, more pixels will contribute to the decision hence loss will go down is my understanding correct?

also what would be the pretrained model argument for this p2 model in `--weights` will it be `yolov5p2.pt` ?

**Comment by glenn-jocher** [comment]
@PushpakBhoge there's simply more for the convolutions to work with if there are more pixels representing the object. Like I said the P2 models are intended for smaller stride outputs down to stride 4.

**Comment by PushpakBhoge** [comment]
ok thank you for all the help! one last thing is there coco pretrained model available for p2 model ?

**Comment by glenn-jocher** [comment]
@PushpakBhoge no, but you can pair any --weights with any --cfg. Any layers with matching names and shapes will transfer.

**Comment by PushpakBhoge512** [comment]
ok thanks that will do thank you so much for the help!

**Comment by Akhp888** [comment]
@glenn-jocher I see Sahi wrapper ( sliding window) on yolov5 being used by many , wanted to know if that impacts the yolov5 results in any way and if thats recommended .

**Comment by glenn-jocher** [comment]
@Akhp888 I'd consult the SAHI author

**Comment by ExtReMLapin** [comment]
@PushpakBhoge have you managed to improve your metrics, and by how much ?

**Comment by PushpakBhoge** [comment]
@ExtReMLapin I have tried with higher resolution training didn't really have much impact in confidence score in my case.
But longer training did help it increased matrices on higher confidence thresholds by thresholds betn 0.3-0.5 I observed improvement of approx 15% while for thresholds Between 0.5 to 0.7 I observed improvement of 40-60%

**Comment by glenn-jocher** [comment]
Yes if confidence scores are low this in an indicator of too-short training epochs.

**Comment by ExtReMLapin** [comment]
Alright, i'll give a try, thank you both of you !

**Comment by CesareDavidePace** [comment]
@ExtReMLapin i have the same problem. did you solve it then?

**Comment by ExtReMLapin** [comment]
I was annotating minutiaes , on big images, so there was two ways of fixing this,  boosting the annotations size by 3 (and dividing by 3 at inference) and doing tiling, which means you don't reduce your big image size, you just subdivise it in 1024x1024 chunks and feed each chunks one by one.

P2 Model didn't really help me there.

**Comment by CesareDavidePace** [comment]
what does it mean boosting the annotations size by 3?

**Comment by ExtReMLapin** [comment]
I had all my annotations and I decided to do : width = width*3, same for height.

**Comment by CesareDavidePace** [comment]
I can't show the dataset, but the problem is that I can't get good confidence in the results. For me, these are large images and I do the tiling 1920x1920. I also sh
![results](https://user-images.githubusercontent.com/104144011/203248754-e3909c60-8a72-4ee0-b223-dfce69649103.png)
are the hyp used.

![results](https://user-images.githubusercontent.com/104144011/203248798-debd29cc-c9b4-4f2f-b737-be038e1c3ec3.png)
![F1_curve](https://user-images.githubusercontent.com/104144011/203248800-f4fe4e7f-e149-4928-b13b-b3f18dc1d91f.png)

lr0: 0.01
lrf: 0.1
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 0.05
cls: 0.3
cls_pw: 1.0
obj: 0.5
obj_pw: 1.0
iou_t: 0.2
anchor_t: 4.0
fl_gamma: 0.0
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 10
translate: 0.1
scale: 1
shear: 3
perspective: 0.0
flipud: 0.0
fliplr: 0.5
mosaic: 1.0
mixup: 0.2
copy_paste: 0.2

I've already tried to train with multiple epochs but the map immediately goes into overfitting. I can't solve this problem even by lowering the obj loss and increasing the data augment

**Comment by RajibChanda** [comment]
@PushpakBhoge  what was your batch size during training? Did you trained with batch size of 1 or some other batch size?

**Comment by CesareDavidePace** [comment]
batch size 4
